ml/data/attack_data_pipeline.py

```
@@ -423,7 +570,96 @@ def _store_attack_data(self, processed_data: Dict[str, Any]):
+            # us count skipped in the same pass.
+            placeholders = ",".join("?" * len(BANDIT_REWARD_OUTCOMES))
+            sql = f"""
+                SELECT outcome, COUNT(*) FROM (
+                    SELECT outcome
+                    FROM attacks
+                    WHERE {where_sql}
+                    ORDER BY timestamp DESC
+                    LIMIT ?
+                )
+                GROUP BY outcome
+            """
```

Ruff (0.15.12)

[error] Possible SQL injection vector through string-based query construction

(S608)